Notify the congestion controller of losses first

ianswett opened the pull request:
Fixes #3495 and #3288 by notifying the congestion controller of lost packets prior to notifying it of acknowledged packets and passing all lost packets into InPersistentCongestion().

The diff is a bit large because I also inlined OnPacketAcked and renamed OnPacketAckedCC to OnPacketAcked.

ghedo commented:
Thanks. I think `AreAllPacketsLost` will still need some clarifications but I haven't been able to find a nice way to word that, so for now this is good.

cpaasch commented:
Looks good to me wrt to issue #3495

ianswett commented:
@martinduke  I think I addressed your comment by commenting that DetectNewlyAckedPackets is responsible for removing packets from sent_packets.  This aligns with the API for DetectLostPackets.  I also changed OnPacketAcked to OnPacketsAcked.

"// Determine which packets have been acknowledged and remove
  // them from sent_packets.
  newly_acked_packets = DetectNewlyAckedPackets(ack, pn_space)"

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -1265,7 +1244,8 @@ OnLossDetectionTimeout():
     // Time threshold loss Detection
-    DetectLostPackets(pn_space)
+    lost_packets = DetectLostPackets(pn_space)
+    OnPacketsLost(lost_packets)
```
```suggestion
    if (!lost_packets.empty()):
      OnPacketsLost(lost_packets)
```

ianswett replied:
If a timeout occurred, then lost_packets should never be empty I believe?  I could add an assert for that if you'd like?

janaiyengar replied:
Hm, that's true. I was thinking about stream cancellations, but that doesn't matter here. assert SGTM

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -1517,7 +1494,7 @@ Invoked from DetectLostPackets when packets are deemed lost.
      CongestionEvent(largest_lost_packet.time_sent)
```
How about also changing this to the following, and removing the line before this (last() is a language-specific method, so I'm suggesting largest())?
```suggestion
     CongestionEvent(lost_packets.largest().time_sent)
```

ianswett replied:
I'm not a big fan of largest here and above.  If you want to make the change, maybe another PR, since this one is already quite large.

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -1108,14 +1108,16 @@ OnAckReceived(ack, pn_space):
     largest_acked_packet[pn_space] =
         max(largest_acked_packet[pn_space], ack.largest_acked)
 
+  // Determine which packets have been acknowledged and remove
```
```suggestion
  // DetectNewlyAckedPackets finds packets have been newly acknowledged and removes
```

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -1108,14 +1108,16 @@ OnAckReceived(ack, pn_space):
 
+  // Determine which packets have been acknowledged and remove
+  // them from sent_packets.
+  newly_acked_packets = DetectNewlyAckedPackets(ack, pn_space)
```
```suggestion
  newly_acked_packets = DetectAndRemoveAckedPackets(ack, pn_space)
```
here and below

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -1128,13 +1130,12 @@ OnAckReceived(ack, pn_space):
-    OnPacketAcked(acked_packet.packet_number, pn_space)
-
-  DetectLostPackets(pn_space)
+  lost_packets = DetectLostPackets(pn_space)
```
```suggestion
  lost_packets = DetectAndRemoveLostPackets(pn_space)
```
here and below

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -1434,26 +1411,27 @@ increases bytes_in_flight.
 ## On Packet Acknowledgement
 
 Invoked from loss detection's OnPacketAcked and is supplied with the
```
```suggestion
Invoked from loss detection's OnAckReceived and is supplied with the
```

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -1499,15 +1477,14 @@ Invoked when an ACK frame with an ECN section is received from the peer.
        max_ack_delay
      congestion_period = pto * kPersistentCongestionThreshold
      // Determine if all packets in the time period before the
      // newest lost packet, including the edges, are marked
```
```suggestion
     // largest newly lost packet, including the edges, are marked
```

ianswett replied:
I'm not sure largest makes this less clearer, at least to me.

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -1434,26 +1411,27 @@ increases bytes_in_flight.
 
 ~~~
-   InCongestionRecovery(sent_time):
+   InLossRecovery(sent_time):
```
```suggestion
   InCongestionRecovery(sent_time):
```
Revert this change. Was this deliberate? The term congestion recovery is used earlier in the spec and the term is defined.

ianswett replied:
This was deliberate.  I've always heard it called loss recovery or just 'the recovery period'

But I can revert it.

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -1434,26 +1411,27 @@ increases bytes_in_flight.
+     for (packet in acked_packets):
+       // Remove from bytes_in_flight.
+       bytes_in_flight -= packet.size
+       if (InLossRecovery(packet.time_sent)):
```
```suggestion
       if (InCongestionRecovery(packet.time_sent)):
```

ianswett commented:
@marten-seemann my reasoning was I thought it was critical to determine what packets were newly acked before invoking loss detection.  In some cases it may be(depending upon details of one's loss detection), but since we update largest_acked_packet[pn_space] outside the code which detects newly acked packets, it's not critical to the pseudocode as it stands today.

That being said, I think the parallelism is nice, so I'd prefer to keep it.

janaiyengar commented:
@marten-seemann : It does change the order -- it does CC increase action after the decrease

ianswett commented:
Assert added, merging.

kazu-yamamoto commented on draft-ietf-quic-recovery.md:
```diff
@@ -1108,14 +1108,18 @@ OnAckReceived(ack, pn_space):
     largest_acked_packet[pn_space] =
         max(largest_acked_packet[pn_space], ack.largest_acked)
 
+  // DetectNewlyAckedPackets finds packets that are newly
```
`DetectNewlyAckedPackets` is not defined in `master`.
